Futures attempt to execute threads directly if those have not started executing

hkaiser opened the pull request:
- adding new API function hpx::threads::get_outer_self_id

Pansysk75 commented:
The latest commit addresses the following race contition on the runs_child_ variable:

Here is `future_data::set_value()`. The modification (resetting) of `runs_child_` happens under a lock. This is alright:
https://github.com/STEllAR-GROUP/hpx/blob/0ad30ef89b666f557de3a91902e745a08a3abe6f/libs/core/futures/include/hpx/futures/detail/future_data.hpp#L559-L561

However, there was a race condition which happened when a scheduled thread was executed very rapidly. Here's the code that schedules a thread:
https://github.com/STEllAR-GROUP/hpx/blob/3236b2fd7c82e56917c0e37d1ecaa92a3097f6fd/libs/core/futures/include/hpx/futures/packaged_continuation.hpp#L233-L237

`this->runs_child_ = spawner(...)` is supposed to keep a reference of the scheduled thread.
When the scheduled thread was executed very quickly, `this->runs_child_ = spawner(...)` could run at the same time (or after) `future_data::set_value()`.
Thus, there was a race condition between `runs_child_.reset()` and `runs_child_ = spawner(...)` . 

Here's the fix I applied:

https://github.com/STEllAR-GROUP/hpx/blob/0ad30ef89b666f557de3a91902e745a08a3abe6f/libs/core/futures/include/hpx/futures/packaged_continuation.hpp#L239-L243

Firstly, I added a lock to prevent concurrent modification of `runs_child_`.
I also added a check for future_data::is_ready(), ensuring we don't keep an unneeded reference to the scheduled thread (in case it has been already executed).

Let's hope this makes things better

Pansysk75 commented:
I also noticed that the code below does not check whether there is a `thread_execution_hint::run_as_child` hint.
 https://github.com/STEllAR-GROUP/hpx/blob/3236b2fd7c82e56917c0e37d1ecaa92a3097f6fd/libs/core/futures/include/hpx/futures/packaged_continuation.hpp#L233-L237
I'm not sure if that was intentional, but my instinct is that we would also want to enable/disable it here depending on the hint.

hkaiser commented:
> I also noticed that the code below does not check whether there is a `thread_execution_hint::run_as_child` hint.
> 
> https://github.com/STEllAR-GROUP/hpx/blob/3236b2fd7c82e56917c0e37d1ecaa92a3097f6fd/libs/core/futures/include/hpx/futures/packaged_continuation.hpp#L233-L237
> 
> 
> I'm not sure if that was intentional, but my instinct is that we would also want to enable/disable it here depending on the hint.

I'd need to look closer, but I believe this is only called after the check has happened before in other places.

Pansysk75 commented:
> 
> I'd need to look closer, but I believe this is only called after the check has happened before in other places.

I did my learning/testing on the cross_pool_injection example. The `.then` would lead to those lines executing, even with no hint placed anywhere.

hkaiser commented:
> Firstly, I added a lock to prevent concurrent modification of `runs_child_`.
> I also added a check for future_data::is_ready(), ensuring we don't keep an unneeded reference to the scheduled thread (in case it has been already executed).

@Pansysk75 please check my commit. I think I was able to remove this lock.

hkaiser commented:
> > > I'd need to look closer, but I believe this is only called after the check has happened before in other places.
> > 
> > 
> > I did my learning/testing on the cross_pool_injection example. The `.then` would lead to those lines executing, even with no hint placed anywhere.
> > Also now that I think about it, the lock could be avoided if the thread was first "registered" (ie created in a suspended state so that we get a thread_id) and then executed/resumed separately. I think I saw that someplace in this PR
> 
> In that case we will need to add the check here as well.

I'm not quite sure what we should do here. If we left this as is we would loose the direct execution optimization in the worst case...

hkaiser commented:
@Pansysk75 I enabled the new feature on gcc10, clang11, and clang12 and now we see hangs: `tests.unit.modules.resource_partitioner.suspend_thread` and `tests.unit.modules.resource_partitioner.suspend_thread_external` hang :/

Pansysk75 commented:
> @Pansysk75 I enabled the new feature on gcc10, clang11, and clang12 and now we see hangs: `tests.unit.modules.resource_partitioner.suspend_thread` and `tests.unit.modules.resource_partitioner.suspend_thread_external` hang :/

I'll have a look

hkaiser commented:
@Pansysk75 FWIW, my last commit has improved things, but not entirely fixed the hangs.

hkaiser commented on libs/core/futures/src/future_data.cpp:
```diff
@@ -131,6 +131,11 @@ namespace hpx::lcos::detail {
+        {
+            return false;
+        }
+
```
Good catch, that also mean we can remove these lines: https://github.com/STEllAR-GROUP/hpx/blob/3a5a47a0a5fcc853971742bf853b92eb16239ded/libs/core/futures/src/future_data.cpp#L120-L123

Pansysk75 replied:
Sure, just left them in the case that taking a thread_id_ref_type was expensive. Give a thumbs up if you think it's ok to change it

hkaiser replied:
I have looked again and I have convinced myself that there is no need to protect the `runs_child_` intrusive pointer needs to be protected in any way. The issue has to be somewhere else... we need to keep looking.

Pansysk75 replied:
Ok, I didn't understand exactly where the issue was supposed to occur, so I'll just take your word for it.

Currently I can barely reproduce the current assert failure (where a tid from terminated_threads_ does not appear in the thread_map_, when it should), and the issue goes away when logging... I'll give it another go, might find a more reliable failure case

hkaiser replied:
I think this assert is wrong to begin with, I'm looking at it as we speak.

hkaiser commented:
@Pansysk75 the next error (segfault) happens in `suspend_pool_external`, however I'm not able to reproduce it locally. Could you try that, please?

Pansysk75 commented:
> @Pansysk75 the next error (segfault) happens in `suspend_pool_external`, however I'm not able to reproduce it locally. Could you try that, please?

Copying the exact setup on Rostam did it for me.
That means doing
`source src/.jenkins/lsu/env-gcc-10.sh`
building with 
`-DHPX_WITH_PARALLEL_TESTS_BIND_NONE=ON -DCMAKE_BUILD_TYPE=Debug -DHPX_WITH_CHECK_MODULE_DEPENDENCIES=ON -DHPX_WITH_PARCELPORT_COUNTERS=ON -DLCI_DEBUG=ON -DHPX_WITH_CXX_STANDARD=17 -DHPX_WITH_MALLOC=system -DHPX_WITH_FETCH_ASIO=ON -DHPX_WITH_ASIO_TAG=asio-1-12-0 -DHPX_WITH_COMPILER_WARNINGS=ON -DHPX_WITH_COMPILER_WARNINGS_AS_ERRORS=ON -DHPX_WITH_PARCELPORT_MPI=ON -DHPX_WITH_PARCELPORT_LCI=ON -DHPX_WITH_FETCH_LCI=ON -DHPX_COROUTINES_WITH_THREAD_SCHEDULE_HINT_RUNS_AS_CHILD=ON`
and executing
`bin/suspend_thread_external_test --hpx:ini=hpx.stacks.use_guard_pages=0 --hpx:bind=none`


Now, the segfault stack trace points at the catch part of the following try-catch block:
https://github.com/STEllAR-GROUP/hpx/blob/09eabc71da8f2a150b2e26efe3d52d4163e49072/libs/core/thread_pools/include/hpx/thread_pools/scheduled_thread_pool_impl.hpp#L314-L362

It seems like `std::thread()` (inside of`add_processing_unit_internal()` call,  L338) occasionally throws a "Resource temporarily unavailable".

I guess we are not handling that properly, as stop_locked(l) (L358) leads to Scheduler::get_thread_count() being called, and it trying to read uninitialized data.

That's all for now, hope it helps!

hkaiser commented:
@Pansysk75 thanks a lot for this analysis. Nw the next question is why does it throw that exception to begin with? There has to be a reason.

Pansysk75 commented:
> @Pansysk75 thanks a lot for this analysis. Now the next question is why does it throw that exception to begin with? There has to be a reason.

I wrote a script that monitors the amount of memory mappings (using the pmap utility).
Looks like std::thread fails when we have more mappings than specified by `/proc/sys/vm/max_map_count` (currently 65530 on Rostam).

Increasing  `/proc/sys/vm/max_map_count` or decreasing the amount of allocations would allow the tests to complete successfully. The failing tests spawn an absurd amount of futures, so it seems sensible to just limit the amount of futures spawned.

hkaiser commented:
> > @Pansysk75 thanks a lot for this analysis. Now the next question is why does it throw that exception to begin with? There has to be a reason.
> 
> I wrote a script that monitors the amount of memory mappings (using the pmap utility). Looks like std::thread fails when we have more mappings than specified by `/proc/sys/vm/max_map_count` (currently 65530 on Rostam).
> 
> Increasing `/proc/sys/vm/max_map_count` or decreasing the amount of allocations would allow the tests to complete successfully. The failing tests spawn an absurd amount of futures, so it seems sensible to just limit the amount of futures spawned.
> 
> (Also `--hpx:ini=hpx.stacks.use_guard_pages=0` results in the accumulation of a higher number of (but smaller in individual size) mappings. But I do not know enough of what's happening to interpret that observation.)

Thanks! I have at least managed to avoid the segfault now and properly report the exception. ;-)

Pansysk75 commented:
@hkaiser Here's what I see happening:

Work started without the hint, gets scheduled through `register_work`, which does not directly create a thread object, but only adds the thread descripition to Scheduler.new_data_. That keeps the amount of allocations low, because the allocated stack memory can be reused. 

Work started with the runs_as_child hint gets handled through `register_thread`, which leads to Scheduler::create_thread always creating the thread object (and allocating the memory for it). So if a bulk of tasks (ie through hpx::async) are created, we immediately do as many stack allocations as there are threads created, and there is no chance for memory recycling to happen

So, the exessive memory usage is not a bug, rather it's a design issue. I remember you telling me about how you wanted to delay the thread allocation, now I get why that could be a good idea :)

hkaiser commented:
> @hkaiser Here's what I see happening:
> 
> Work started without the runs_as_child hint, gets scheduled through `register_work`, which does not directly create a thread object, but only adds the thread descripition to Scheduler.new_data_. The actual thread will be allocated/created when the Scheduler starts running out of threads in the queues. That keeps the amount of allocations low, because, since the threads are created on an as-needed basis, the allocated thread-stack memory can be reused.
> 
> Work started with the runs_as_child hint gets handled through `register_thread`, which leads to Scheduler::create_thread always creating the thread object (and allocating the memory for it). So if a bulk of tasks are created (ie through a ton of calls of hpx::async), we immediately do as many stack allocations as there are threads created, and there is no chance for memory recycling to happen
> 
> So, the exessive memory usage is not a bug, rather it's a design issue. I remember you telling me about how you wanted to delay the thread allocation, now I get why that could be a good idea :)

Yes, that's it! Very nice analysis! The creation of a thread with the hint uses `register_thread` instead of `register_work` as we need the thread id. A possible solution would be (as you said) to recycle the stacks separately from the thread objects, possibly we can even avoid recycling the thread objects themselves, just the stacks. Another solution could be to delay storing the `thread_id` until the thread is actually created.

Pansysk75 commented:
Although disabling run_as_child decreases allocations by a lot, here is still a memory-leak-like issue arising with hpx:ini=hpx.stacks.use_guard_pages=0. I'm investigating it rn.

hkaiser commented:
That would explain why the suspend_runtime et.al. now fail even for builders that don't enable the run_as_child hint by default.

Pansysk75 commented:
> That would explain why the suspend_runtime et.al. now fail even for builders that don't enable the run_as_child hint by default.

Here, does mmap-ing memory of (size + EXEC_PAGESIZE)  make sense even if we are not using guard pages?

https://github.com/STEllAR-GROUP/hpx/blob/dd17010adf02cdab681d96f10f96626d476459c0/libs/core/coroutines/include/hpx/coroutines/detail/posix_utility.hpp#L84-L96

Deallocation seems to not take EXEC_PAGESIZE into account if we don't use guard pages:

https://github.com/STEllAR-GROUP/hpx/blob/dd17010adf02cdab681d96f10f96626d476459c0/libs/core/coroutines/include/hpx/coroutines/detail/posix_utility.hpp#L156-L172

hkaiser commented:
Excellent catch! Could you patch that, please?

Pansysk75 commented:
> Excellent catch! Could you patch that, please?

Thank you, I'll do that :)

hkaiser commented:
I'm also preparing a change right now that starts deleting stack segments of recycled threads once more then a certain number of threads have been recycled. That should keep the amount of memory allocated down a bit as well.

hkaiser commented:
@Pansysk75 your fix did the trick! I will roll back my change that limits the numer of recycled stacks as it made the performance tests fail. Let's see if after that change things will be still green.

Pansysk75 commented:
> @Pansysk75 your fix did the trick! I will roll back my change that limits the numer of recycled stacks as it made the performance tests fail. Let's see if after that change things will be still green.

Yes, things are much better.

The memory mappings now barely go above a few hundred, as the OS is able to merge the mapped areas more efficiently. Even with guard_pages=on, they are in the hundreds now (an order of magnitude less than before).
That seems very significant, so keep an eye out for any macroscopic effects it may have.

The only memory "concern" is the slightly higher memory usage when using direct execution (because of the immediate thread spawining), but it's not that significant. So the thread stack allocation / recycling change is not as crucial.

hkaiser commented:
@Pansysk75 I think this is finally good to be merged. What do you think?

hkaiser commented:
> @hkaiser There is a single timed-out test on daint (I had a look yesterday but didn't get anywhere), but other than that, nothing further to add.

I believe this is unrelated. Could be another case of the synchronization object going out of scope too early.

hkaiser commented:
bors merge
